[IMP] charts: highlights elements on hover

anhe-odoo opened the pull request:
## Task Description

This PR aims to add some interactivity within the charts to allow the user to highlights specific elements in chart when hovering related elements (legend items, pie parts, ...). This also fix the legend behavior to be able to hide/show elements on click.

## Related Task(s)
- Task: [3782657](https://www.odoo.com/odoo/project/2328/tasks/3782657)

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/3784).

hokolomopo commented on src/helpers/figures/charts/line_chart.ts:
```diff
@@ -209,6 +212,56 @@ export class LineChart extends AbstractChart {
+    dataSets.forEach((dataset) => {
+      const backgroundColors = dataset.backgroundColor;
+      backgroundColors.forEach((color, i, colors) => {
+        colors[i] = setColorAlpha(color, 1);
```
maybe use lightenColor rather than setting the alpha, so the behavior is the same for all chart backgroud colors

hokolomopo commented on src/helpers/figures/charts/bar_chart.ts:
```diff
@@ -231,7 +232,7 @@ export function createBarChartRuntime(chart: BarChart, getters: Getters): BarCha
+      ...getChartDefaultOptions("bar"),
       indexAxis: chart.horizontal ? "y" : "x",
       layout: getChartLayout(definition, chartData),
       scales: getBarChartScales(definition, chartData),
```
add `onHover: getBarChartOnHover(...)` methods to stay consistend with the rest, rather than a `getChartDefaultOptions`

hokolomopo commented on src/helpers/figures/charts/bar_chart.ts:
```diff
@@ -242,6 +243,7 @@ export function createBarChartRuntime(chart: BarChart, getters: Getters): BarCha
         chartShowValuesPlugin: getChartShowValues(definition, chartData),
       },
     },
+    plugins: [getChartMouseOutPlugin("bar")],
```
same thing here, create helper `getBarChartMouseOutPlugin`

hokolomopo commented on packages/o-spreadsheet-engine/src/helpers/figures/charts/runtime/chart_highlight.ts:
```diff
@@ -0,0 +1,123 @@
+        unHighlightItems(c.data.datasets);
+        c.update();
+      }
+    },
```
everything's typed as any >_<

hokolomopo commented on packages/o-spreadsheet-engine/src/helpers/figures/charts/runtime/chart_highlight.ts:
```diff
@@ -0,0 +1,123 @@
+
+function highlightPieChartItem(item: { index?: number }, dataSets: ChartDataset[]) {
+  const index = item.index;
+  dataSets.forEach((dataset) => {
```
```suggestion
  for(const dataset of datasets)```

some people don't like foreach in here

hokolomopo commented on packages/o-spreadsheet-engine/src/helpers/figures/charts/runtime/chart_highlight.ts:
```diff
@@ -0,0 +1,123 @@
+function highlightLineChartItem(item: { datasetIndex?: number }, dataSets: ChartDataset[]) {
+  const index = item.datasetIndex;
+  dataSets.forEach((dataset, i) => {
+    const color = setColorAlpha(dataset.borderColor as Color, i === index ? 1 : 0.4);
```
ooh I think I remember one of the reason I had a hard time with sunburst. `dataset.borderColor` can be a callback, in which case this doesn't work. I don't think we ever use callbacks for pie/bar/line so it should be good here

hokolomopo reviewed (commented):
👋 

Also added testing notes in the pad :)

fdamhaut reviewed (commented):
small nitpick

pro-odoo commented on packages/o-spreadsheet-engine/src/helpers/figures/charts/runtime/chart_highlight.ts:
```diff
@@ -0,0 +1,118 @@
+      return;
+    }
+    backgroundColors.forEach((color, i, colors) => {
+      colors[i] = setColorAlpha(color, i === index ? 1 : 0.4);
```
```suggestion
      colors[i] = setColorAlpha(color, i === item.index ? 1 : 0.4);
```
Useless variable assignation

pro-odoo reviewed (commented):
<img width="237" height="181" alt="Image" src="https://github.com/user-attachments/assets/336355e4-54d0-43ad-b283-f93731f18569" />

anhe-odoo commented on src/helpers/figures/charts/runtime/chartjs_legend.ts:
```diff
@@ -229,46 +291,65 @@ export function getSunburstChartLegend(
+      if (!target || !(target instanceof HTMLElement)) {
+        return;
+      }
+      target.style.cursor = "pointer";
```
Yes, there is a problem with the cursor style in odoo chart in general. This will be fixed with this task : https://www.odoo.com/odoo/project/2328/tasks/5012198

anhe-odoo committed:
[IMP] charts: highlights elements on hover

Task Description

This PR aims to add some interactivity within the charts to allow
the user to highlights specific elements in chart when hovering
related elements (legend items, pie parts, ...). This also fix the
legend bahavior to be able to hide/show elements on click.

Related Task(s)

Task: 3782657

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_dataset.ts:
```diff
@@ -300,6 +300,7 @@ export function getPieChartDatasets(
       borderColor: definition.background || "#FFFFFF",
       backgroundColor,
       hoverOffset: 10,
+      hoverBackgroundColor: backgroundColor,
```
Why this change ? And why only in pie charts ?

anhe-odoo replied:
It's an old relicate of when I was also highlighting the pie part when hovering the chart. I removed then :)

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_legend.ts:
```diff
@@ -228,46 +290,65 @@ export function getSunburstChartLegend(
+      if (!target || !(target instanceof HTMLElement)) {
+        return;
+      }
+      target.style.cursor = "pointer";
```
```suggestion
      if (target && target instanceof HTMLElement) {
      target.style.cursor = "pointer";
      }
```
nitpick (same in onLeave)

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_legend.ts:
```diff
@@ -228,46 +290,65 @@ export function getSunburstChartLegend(
+      }
+      if (!legend.legendItems) {
+        return;
+      }
```
```suggestion
      if (event.type !== "click" || !legend.legendItems) {
        return;
      }
```
nitpick

hokolomopo commented on src/helpers/figures/charts/runtime/chart_highlight.ts:
```diff
@@ -0,0 +1,126 @@
+import { HIGHLIGHT_TRANSPARENCY, LINE_FILL_TRANSPARENCY } from "../../../../constants";
+import { setColorAlpha } from "../../../color";
+
+export function highlightBarChartItem(item: LegendItem, dataSets: ChartDataset<"bar">[]) {
```
this does not seem to work with trend lines

hokolomopo commented on src/helpers/figures/charts/runtime/chart_highlight.ts:
```diff
@@ -0,0 +1,126 @@
+  }
+}
+
+export function highlightComboChartItem(
```
combo/line/bar highlight and unhilight function are 90% the same things. Can they be combined in a single function ? If combo can handle both bar & line datasets, it could be used in both bar & line charts no ?

anhe-odoo replied:
We can't have the line and combo/bar in the same method as we have to apply a different transparency for the line chart background (for the area chart). Or I should start by reading the current opacity, apply a factor to reduce it and then use the `setColorAlpha` with this new value. But I'm not sure of the process then

hokolomopo replied:
Nevermind if it's too annoying

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_legend.ts:
```diff
@@ -77,6 +120,11 @@ export function getPieChartLegend(
+      highlightPieChartItem,
+      resetPieChartHighlights,
+      togglePieDataVisibility
+    ),
```
nitpick: i'd make named arguments, it's a bit more readable IMO
```suggestion
    ...getInteractiveLegendConfig(
      {
      highlightDatasets: highlightPieChartItem,
      unHilightDatasets: resetPieChartHighlights,
      onLegentItemClick: togglePieDataVisibility
      }
    ),
```

hokolomopo reviewed (commented):
Small issue with trendlines 🙃 

Feel free to ignore the nitpicks

need a rebase + prettier. Also the runbot looks broken

hokolomopo commented on tests/figures/chart/pie_chart_plugin.test.ts:
```diff
@@ -65,6 +65,8 @@ describe("pie chart", () => {
         strokeStyle: "#4EA7F2",
         fontColor: "#FFFFFF",
+        index: 0,
+        hidden: true,
```
wait why are the labels hidden by default ?

anhe-odoo replied:
It was caused by a missing method in fakeChart

hokolomopo commented on src/constants.ts:
```diff
@@ -231,6 +231,8 @@ export const DEFAULT_SCORECARD_KEY_VALUE_FONT_SIZE = 32;
 
 export const LINE_FILL_TRANSPARENCY = 0.4;
+export const HIGHLIGHT_TRANSPARENCY = 0.2;
+export const DEFAULT_PIE_BORDER = 2;
```
never used

hokolomopo commented on src/constants.ts:
```diff
@@ -231,6 +231,8 @@ export const DEFAULT_SCORECARD_KEY_VALUE_FONT_SIZE = 32;
 export const DEFAULT_SCORECARD_BASELINE_FONT_SIZE = 16;
 
 export const LINE_FILL_TRANSPARENCY = 0.4;
+export const HIGHLIGHT_TRANSPARENCY = 0.2;
```
either define it in the chart_highlight file (since it's only used here), or rename it to `chart_something` if you leave it in the global constant file

hokolomopo reviewed (commented):
👋 

Small details, otherwise LGTM :) 

I don't think the odoo PR is needed anymore ? But still please still rebase the odoo branch, so we can test that it works correctly for odoo charts

hokolomopo commented:
@robodoo r+
